File: boilerplate/java/order-service/src/main/java/com/example/orderservice/infrastructure/cache/RedisCacheManager.java
package com.example.orderservice.infrastructure.cache;

import com.example.orderservice.domain.ports.CacheException;
import com.example.orderservice.domain.ports.CacheManager;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Component;

import java.time.Duration;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.TimeUnit;

/**
 * Redis implementation of CacheManager.
 * 
 * Uses Spring Data Redis for cache operations.
 * Default TTL: 30 minutes for most cache entries.
 */
@Component
public class RedisCacheManager implements CacheManager {
    
    private static final Duration DEFAULT_TTL = Duration.ofMinutes(30);
    
    private final RedisTemplate<String, Object> redisTemplate;
    
    public RedisCacheManager(RedisTemplate<String, Object> redisTemplate) {
        this.redisTemplate = redisTemplate;
    }
    
    @Override
    @SuppressWarnings("unchecked")
    public <T> Optional<T> get(String key, Class<T> type) {
        try {
            Object value = redisTemplate.opsForValue().get(key);
            if (value == null) {
                return Optional.empty();
            }
            if (type.isInstance(value)) {
                return Optional.of((T) value);
            } else {
                throw new CacheException(
                    "Cached value type mismatch. Expected: " + type.getName() + 
                    ", Got: " + (value != null ? value.getClass().getName() : "null")
                );
            }
        } catch (Exception e) {
            throw new CacheException("Failed to get cache key: " + key, e);
        }
    }
    
    @Override
    public <T> void put(String key, T value) {
        put(key, value, DEFAULT_TTL);
    }
    
    @Override
    public <T> void put(String key, T value, Duration ttl) {
        try {
            redisTemplate.opsForValue().set(key, value, ttl.toMillis(), TimeUnit.MILLISECONDS);
        } catch (Exception e) {
            throw new CacheException("Failed to put cache key: " + key, e);
        }
    }
    
    @Override
    public void evict(String key) {
        try {
            Boolean deleted = redisTemplate.delete(key);
            if (Boolean.FALSE.equals(deleted)) {
                // Key didn't exist - this is OK, no-op
            }
        } catch (Exception e) {
            throw new CacheException("Failed to evict cache key: " + key, e);
        }
    }
    
    @Override
    public boolean contains(String key) {
        try {
            return Boolean.TRUE.equals(redisTemplate.hasKey(key));
        } catch (Exception e) {
            throw new CacheException("Failed to check cache key existence: " + key, e);
        }
    }
    
    @Override
    public void clearPattern(String pattern) {
        try {
            Set<String> keys = redisTemplate.keys(pattern);
            if (keys != null && !keys.isEmpty()) {
                redisTemplate.delete(keys);
            }
        } catch (Exception e) {
            throw new CacheException("Failed to clear cache pattern: " + pattern, e);
        }
    }
    
    @Override
    public void clearAll() {
        try {
            Set<String> keys = redisTemplate.keys("*");
            if (keys != null && !keys.isEmpty()) {
                redisTemplate.delete(keys);
            }
        } catch (Exception e) {
            throw new CacheException("Failed to clear all cache", e);
        }
    }
}

File: tools/cookiecutter-python/{{cookiecutter.project_slug}}/src/infrastructure/cache/redis_cache_manager.py
"""
Redis implementation of CacheManager.

Uses redis-py (async) for cache operations.
Default TTL: 30 minutes for most cache entries.
"""

import json
from typing import Any, Optional, Type
from datetime import timedelta
from domain.ports.cache_manager import CacheManager, CacheException


class RedisCacheManager(CacheManager):
    """Redis-based cache manager implementation."""
    
    DEFAULT_TTL = timedelta(minutes=30)
    
    def __init__(self, redis_client: Any):
        """
        Initialize Redis cache manager.
        
        Args:
            redis_client: Redis async client instance
        """
        self.redis = redis_client
    
    def get(self, key: str, type_: Type) -> Optional[Any]:
        """Get value from cache."""
        try:
            value = self.redis.get(key)
            if value is None:
                return None
            
            # Deserialize JSON
            deserialized = json.loads(value)
            
            # Type checking is best-effort in Python
            return deserialized
        except Exception as e:
            raise CacheException(f"Failed to get cache key: {key}") from e
    
    def put(self, key: str, value: Any, ttl: Optional[timedelta] = None) -> None:
        """Put value in cache with optional TTL."""
        try:
            ttl_seconds = int((ttl or self.DEFAULT_TTL).total_seconds())
            serialized = json.dumps(value)
            self.redis.set(key, serialized, ex=ttl_seconds)
        except Exception as e:
            raise CacheException(f"Failed to put cache key: {key}") from e
    
    def evict(self, key: str) -> None:
        """Remove value from cache."""
        try:
            self.redis.delete(key)
        except Exception as e:
            raise CacheException(f"Failed to evict cache key: {key}") from e
    
    def contains(self, key: str) -> bool:
        """Check if key exists in cache."""
        try:
            return self.redis.exists(key) > 0
        except Exception as e:
            raise CacheException(f"Failed to check cache key existence: {key}") from e
    
    def clear_pattern(self, pattern: str) -> None:
        """Clear all cache entries matching pattern."""
        try:
            keys = self.redis.keys(pattern)
            if keys:
                self.redis.delete(*keys)
        except Exception as e:
            raise CacheException(f"Failed to clear cache pattern: {pattern}") from e
    
    def clear_all(self) -> None:
        """Clear entire cache."""
        try:
            self.redis.flushdb()
        except Exception as e:
            raise CacheException("Failed to clear all cache") from e


def create_redis_client(host: str = "redis", port: int = 6379, db: int = 0) -> Any:
    """
    Create Redis client instance.
    
    Args:
        host: Redis hostname (default: "redis" for Docker network)
        port: Redis port (default: 6379)
        db: Redis database number (default: 0)
        
    Returns:
        Redis client instance
    """
    import redis
    
    return redis.Redis(
        host=host,
        port=port,
        db=db,
        decode_responses=True,
        socket_connect_timeout=5,
        socket_timeout=5,
    )
